design-documents/service-isolation/cicd-pipeline.md

```diff
@@ -0,0 +1,51 @@
+
+In the scope of service isolation we want to have CI/CD process that would allow
+* Running tests on change set and update preview environment
+* PR delivery to mainline
+
+And in addition to that support the following scenarios that can be used internally and by system integrators
+* Update stage environment
+* Update production environment
+* Support blue/green deployment
+
+Steps that are currently not automated
+* Creating builds on git push
+* Updating preview environment (teams use for testing)
+* Updating stage environment (teams use for demo purposes)
+* PR merge
+* Updating production environment and blue/green deployment
+
+Pipelines should be configured by code and/or convention. Triggered by pushing code. There are should be minimal/or no interaction with UI (with the exception of PR approvals) if the pipeline is successful.
+
+## Design
+
+Branches named by convention. Branches that start with feature contain clean feature related changes. Pushing changes to feature branches triggers CI pipeline below. User receive notification via slack/email with links to builds.
+
+![New CI pipeline](img/new-ci-pipeline.png)
+
+Branches with delivery prefix contains changes that will be delivered in the sprint to mainline, pushing changes to these branches will start the following pipeline. User will receive similar notifications.
+
+![New CI pipeline](img/new-ci-pipeline-merge-pr.png)
+
+### Dealing with multiple repos/branches
+
+We need to account for the case when developer makes a change in one repo, but we need to create a build for multiple repos. Possible solutions are below:
+
+Multiple repos
+* Have configuration of what repos need to be included in the build, this usually doesn't change, this can help solve problem with repos. File with this information can be put in each repo or in the repo with pipeline configuration.
+
+Multiple branches
+* Commit a file that will contain configuration to orchestrate pipeline with information about branches in other repos that need to be included in the build. Each repo may contain default file, which would contain configuration to look for branches by convention and default to master. Configuration can be overwritten if needed. This also solves issue with multiple repos. File with overwritten configuration will be removed before merge.
+* Search for a branches by convention
+* Commit configuration for a build in a separate repo that would serve a purpose of orchestrating pipeline. Similar to #1, but file committed in a different repo.
+
+## Restarting pipeline
+Pipeline can be created for each commit and restarted if necessary. CI should send links to pipeline via slack or email.
+
+## Open questions
+* Do we want to choose technology that is build on Tekton?
+* How to not start unnecessary pipelines when pushing changes to different repos? For instance, after pushing changes to repo A, want to push changes to repo B. Should we cancel previous builds? Will cancelling previous builds always be ok?
```

Comment:
Also, how it will work with force push? Should we have a grace period after the push. Maybe we should stop builds if commit message has `Fix test` in it?